src/devices/bus/s100/vectordualmode.cpp

```diff
@@ -0,0 +1,315 @@
+
+/* Interleave 8 bits with zeros. abcdefgh -> 0a0b0c0d0e0f0g0h */
+static int deposit8(int data)
+{
+	int d = data;
+	d = ((d & 0xf0) << 4) | (d & 0x0f);
+	d = ((d << 2) | d) & 0x3333;
+	d = ((d << 1) | d) & 0x5555;
+	return d;
+}
+
+static uint16_t mfm_byte(uint8_t data, unsigned int prev_data)
+{
+	const unsigned int ext_data = data | (prev_data << 8);
+	const unsigned int clock = ~(ext_data | (ext_data >> 1));
+	return (deposit8(clock) << 1) | deposit8(ext_data);
+}
+
+static uint8_t unmfm_byte(uint16_t mfm)
+{
+	unsigned int d = mfm;
+	d &= 0x5555;
+	d = ((d >> 1) | d) & 0x3333;
+	d = ((d >> 2) | d) & 0x0f0f;
+	d = ((d >> 4) | d) & 0x00ff;
+	return d;
+}
+
+s100_vector_dualmode_device::s100_vector_dualmode_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, S100_VECTOR_DUALMODE, tag, owner, clock)
+	, device_s100_card_interface(mconfig, *this)
+	, m_floppy(*this, "floppy%u", 0U)
+	, m_ram{0}
+	, m_cmar(0)
+	, m_drive(0)
+	, m_sector(0)
+	, m_read(false)
+	, m_pll()
+	, m_byte_timer(nullptr)
+	, m_sector_timer(nullptr)
+	, m_pending_byte(0)
+	, m_pending_size(0)
+{
+}
+
+TIMER_CALLBACK_MEMBER(s100_vector_dualmode_device::motor_off)
+{
+	for (int i = 0; i < m_floppy.size(); i++) {
+		floppy_image_device* flop = m_floppy[m_drive]->get_device();
+		if (flop)
+			flop->mon_w(1);
+	}
+}
+
+bool s100_vector_dualmode_device::hdd_selected()
+{
+	return m_drive == 0 && false;
+}
+
+uint8_t s100_vector_dualmode_device::s100_sinp_r(offs_t offset)
+{
+	// 7200-1200-02-1 page 16 (1-10)
+	uint8_t data;
+	if (offset == 0xc0) { // status (0) port
+		bool write_protect; // FDD
+		bool ready; // HDD
+		bool track0;
+		bool write_fault = false; // HDD
+		bool seek_complete; // HDD
+		bool loss_of_sync; // HDD
+		if (hdd_selected()) {
+			write_protect = false;
+			ready = true;
+			track0 = false;
+			seek_complete = true;
+			loss_of_sync = true;
+		} else {
+			floppy_image_device* flop = m_floppy[m_drive]->get_device();
+			write_protect = flop && flop->wpt_r();
+			ready = false;
+			track0 = flop && !flop->trk00_r();
+			seek_complete = false;
+			loss_of_sync = false;
+		}
+
+		data = write_protect | (ready << 1) | (track0 << 2)
+		    | (write_fault << 3) | (seek_complete << 4) | (loss_of_sync << 5)
+			| 0xc0;
+	} else if (offset == 0xc1) { // status (1) port
+		bool floppy_disk_selected;
+		bool controller_busy = m_sector_timer->enabled();
+		bool motor_on; // FDD
+		bool type_of_hard_disk = true;
+		if (hdd_selected()) {
+			floppy_disk_selected = false;
+			motor_on = false;
+		} else {
+			floppy_disk_selected = true;
+			motor_on = m_motor_on_timer->enabled();
+		}
+		data = floppy_disk_selected | (controller_busy << 1) | (motor_on << 2)
+		    | (type_of_hard_disk << 3) | 0xf0;
+	} else if (offset == 0xc2) { // data port
+		data = m_ram[m_cmar++];
+		m_cmar &= 0x1ff;
+	} else if (offset == 0xc3) { // reset port
+		m_cmar = 0;
+		data = 0xff;
+	} else {
```

Comment:
You need to guard against debugger accesses on reads with side effects.  You need to check `if (!machine().side_effects_disabled())` before modifying contents of `m_cmar` here or bad things can happen when a debugger memory window is open.